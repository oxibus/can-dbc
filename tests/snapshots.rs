#![cfg(all(feature = "serde", feature = "encodings"))]

use nom::Err;
use std::borrow::Cow;
use std::fs;
use std::path::{Path, PathBuf};

use can_dbc::{decode_cp1252, Dbc, Error};
use insta::{assert_debug_snapshot, assert_yaml_snapshot, with_settings};
use test_each_file::test_each_path;

struct TestConfig {
    test_root: &'static str,
    snapshot_suffix: &'static str,
    use_cp1251: bool,
    create_snapshot: bool,
}

static TEST_DIRS: &[TestConfig] = &[
    TestConfig {
        test_root: "shared-test-files",
        snapshot_suffix: "",
        use_cp1251: true,
        create_snapshot: true,
    },
    TestConfig {
        test_root: "opendbc/opendbc/dbc",
        snapshot_suffix: "opendbc",
        use_cp1251: false,
        create_snapshot: false,
    },
];

test_each_path! { for ["dbc"] in "./tests/fixtures/opendbc/opendbc/dbc" as dbc => parse_one_file }
test_each_path! { for ["dbc"] in "./tests/fixtures/shared-test-files" as shared => parse_one_file }
// upper case extension
test_each_path! { for ["DBC"] in "./tests/fixtures/shared-test-files" as shared2 => parse_one_file }

/// Get `test root`, `snapshot name suffix`, `use cp1251`, `create snapshot` for the given path
fn get_test_info(path: &Path) -> Option<(PathBuf, &'static TestConfig)> {
    if !path
        .extension()
        .unwrap_or_default()
        .eq_ignore_ascii_case("dbc")
    {
        return None;
    }
    let path_str = path.to_str().unwrap();
    let parent = path.parent().unwrap();
    for item in TEST_DIRS {
        // Ensure slashes are there for easier matching
        let test_root = format!("/{}/", item.test_root);
        let mut path_dir = parent.to_str().unwrap().to_string();
        if !path_dir.ends_with('/') {
            path_dir.push('/');
        }
        if let Some(pos) = path_dir.find(&test_root) {
            let parent = PathBuf::from("snapshots")
                .join(item.snapshot_suffix)
                .join(&path_dir[pos + test_root.len()..]);
            return Some((parent, item));
        }
    }
    panic!("Unknown test directory: {path_str}");
}

/// Test parsing all DBC files
#[test]
fn test_if_submodules_are_present() {
    for test in TEST_DIRS {
        let dir = Path::new("./tests/fixtures").join(test.test_root);
        fs::read_dir(&dir).unwrap_or_else(|e| {
            let dir_display = dir.display();
            panic!(
                "
--------------------------------------------------------------------------
Error reading dbc test files from   {dir_display}
{e}
Make sure git submodules are up to date by running
    git submodule update --init --recursive
--------------------------------------------------------------------------
"
            )
        });
    }
}

/// Parse a single DBC file and assert a snapshot of the result.
fn parse_one_file([path]: [&Path; 1]) {
    let Some((
        snapshot_path,
        &TestConfig {
            use_cp1251,
            create_snapshot,
            ..
        },
    )) = get_test_info(path)
    else {
        return;
    };

    let file_name = path.file_stem().unwrap().to_string_lossy().to_string();
    let buffer = fs::read(path).unwrap();
    let buffer = if use_cp1251 {
        decode_cp1252(&buffer)
            .unwrap_or_else(|| panic!("Failed to decode {} as cp1252", path.display()))
    } else {
        Cow::Borrowed(
            std::str::from_utf8(&buffer)
                .unwrap_or_else(|_| panic!("Failed to decode {} as utf-8", path.display())),
        )
    };

    report_test_result(
        Dbc::try_from(buffer.as_ref()),
        snapshot_path,
        file_name,
        create_snapshot,
    );
}

fn report_test_result(
    result: Result<Dbc, Error>,
    snapshot_path: PathBuf,
    file_name: String,
    do_snapshot: bool,
) {
    let do_snapshot = std::env::var("SKIP_INSTA").is_err()
        && (do_snapshot || std::env::var("FORCE_INSTA").is_ok());

    with_settings! {
        {
            omit_expression => true,
            snapshot_path => snapshot_path,
            prepend_module_to_snapshot => false
        },
        {
            match result {
                Ok(result) => {
                    if do_snapshot {
                        assert_yaml_snapshot!(file_name, result);
                    } // else ignore successful results when not snapshotting
                }
                Err(e) => {
                    let file_name = format!("!error___{file_name}");
                    assert_debug_snapshot!(file_name, err_to_str(&e));
                }
            }
        }
    }
}

fn err_to_str<'a>(err: &'a Error) -> &'a str {
    match err {
        Error::Incomplete(_, _) => "incomplete",
        Error::Nom(e) => match e {
            Err::Incomplete(_) => "nom-incomplete",
            Err::Error(_) => "nom-error",
            Err::Failure(_) => "nom-failure",
        },
        Error::MultipleMultiplexors => "multiple multiplexors",
    }
}
