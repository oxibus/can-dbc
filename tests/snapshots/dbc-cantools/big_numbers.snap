---
source: tests/snapshots.rs
---
version: ""
new_symbols:
  - NS_DESC_
  - CM_
  - BA_DEF_
  - BA_
  - VAL_
  - CAT_DEF_
  - CAT_
  - FILTER
  - BA_DEF_DEF_
  - EV_DATA_
  - ENVVAR_DATA_
  - SGTYPE_
  - SGTYPE_VAL_
  - BA_DEF_SGTYPE_
  - BA_SGTYPE_
  - SIG_TYPE_REF_
  - VAL_TABLE_
  - SIG_GROUP_
  - SIG_VALTYPE_
  - SIGTYPE_VALTYPE_
  - BO_TX_BU_
  - BA_DEF_REL_
  - BA_REL_
  - BA_DEF_DEF_REL_
  - BU_SG_REL_
  - BU_EV_REL_
  - BU_BO_REL_
  - SG_MUL_VAL_
bit_timing: []
nodes:
  - TheNode
messages:
  - id:
      Extended: 57
    name: TheMessage
    size: 8
    transmitter: VectorXXX
    signals:
      - name: TheSignal
        multiplexer_indicator: Plain
        start_bit: 0
        size: 8
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
  - id:
      Standard: 42
    name: TheOtherMessage
    size: 8
    transmitter: VectorXXX
    signals:
      - name: TheSignal
        multiplexer_indicator: Plain
        start_bit: 0
        size: 8
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
comments:
  - Node:
      name: TheNode
      comment: TheNodeComment
attribute_definitions:
  - Signal: "\"TheSignalStringAttribute\" STRING"
  - Plain: "\"TheNetworkAttribute\" INT 0 100"
  - Message: "\"TheUnusedAttributeDefinitionWithoutDefault\" HEX 0 8"
  - Message: "\"TheHexAttribute\" HEX 0 8"
  - Message: "\"TheFloatAttribute\" FLOAT 5 87"
  - Node: "\"TheNodeAttribute\" INT -9.22337203685478E+018 1.84467440737096E+019"
  - Node: "\"NodeLayerModules\" STRING"
  - Message: "\"GenMsgStartDelayTime\" INT 0 65535"
  - Signal: "\"NWM-WakeupAllowed\" ENUM  \"No\",\"Yes\""
  - Message: "\"NmMessage\" ENUM  \"no\",\"yes\""
  - Message: "\"GenMsgILSupport\" ENUM  \"No\",\"Yes\""
  - Node: "\"NmNode\" ENUM  \"no\",\"yes\""
  - Node: "\"NmStationAddress\" INT 0 63"
  - Plain: "\"NmBaseAddress\" HEX 1024 1087"
  - Message: "\"GenMsgCycleTimeFast\" INT 0 50000"
  - Plain: "\"BusType\" STRING"
  - Message: "\"GenMsgCycleTime\" INT 0 50000"
  - Message: "\"GenMsgDelayTime\" INT 0 1000"
  - Message: "\"GenMsgNrOfRepetition\" INT 0 999999"
  - Message: "\"GenMsgSendType\" ENUM  \"Cyclic\",\"NotUsed\",\"NotUsed\",\"NotUsed\",\"NotUsed\",\"NotUsed\",\"NotUsed\",\"IfActive\",\"NoMsgSendType\",\"NotUsed\",\"vector_leerstring\""
  - Signal: "\"GenSigInactiveValue\" INT 0 100000"
  - Signal: "\"GenSigSendType\" ENUM  \"Cyclic\",\"OnWrite\",\"OnWriteWithRepetition\",\"OnChange\",\"OnChangeWithRepetition\",\"IfActive\",\"IfActiveWithRepetition\",\"NoSigSendType\",\"NotUsed\",\"NotUsed\",\"NotUsed\",\"NotUsed\",\"NotUsed\""
  - Signal: "\"GenSigStartValue\" FLOAT 0 100000000000"
attribute_defaults:
  - name: TheSignalStringAttribute
    value:
      String: "100"
  - name: TheNetworkAttribute
    value:
      Double: 50
  - name: TheHexAttribute
    value:
      Double: 4
  - name: TheFloatAttribute
    value:
      Double: 55
  - name: TheNodeAttribute
    value:
      Double: 100
  - name: NodeLayerModules
    value:
      String: CANoeILNLVector.dll
  - name: GenMsgStartDelayTime
    value:
      Double: 0
  - name: NWM-WakeupAllowed
    value:
      String: ""
  - name: NmMessage
    value:
      String: "no"
  - name: GenMsgILSupport
    value:
      String: "Yes"
  - name: NmNode
    value:
      String: "no"
  - name: NmStationAddress
    value:
      Double: 0
  - name: NmBaseAddress
    value:
      Double: 1024
  - name: GenMsgCycleTimeFast
    value:
      Double: 0
  - name: BusType
    value:
      String: ""
  - name: GenMsgCycleTime
    value:
      Double: 0
  - name: GenMsgDelayTime
    value:
      Double: 0
  - name: GenMsgNrOfRepetition
    value:
      Double: 0
  - name: GenMsgSendType
    value:
      String: NoMsgSendType
  - name: GenSigInactiveValue
    value:
      Double: 0
  - name: GenSigSendType
    value:
      String: Cyclic
  - name: GenSigStartValue
    value:
      Double: 0
attribute_values_database:
  - name: TheNetworkAttribute
    value:
      Double: 51
  - name: BusType
    value:
      String: CAN
attribute_values_node:
  - name: TheNodeAttribute
    node_name: TheNode
    value:
      Double: 99
attribute_values_message:
  - name: TheFloatAttribute
    message_id:
      Extended: 57
    value:
      Double: 58.7
  - name: TheHexAttribute
    message_id:
      Extended: 57
    value:
      Double: 5
  - name: GenMsgSendType
    message_id:
      Extended: 57
    value:
      Double: 0
  - name: GenMsgCycleTime
    message_id:
      Extended: 57
    value:
      Double: 1000
attribute_values_signal:
  - name: TheSignalStringAttribute
    message_id:
      Extended: 57
    signal_name: TheSignal
    value:
      String: TestString
  - name: GenSigSendType
    message_id:
      Extended: 57
    signal_name: TheSignal
    value:
      Double: 1
  - name: GenSigSendType
    message_id:
      Standard: 42
    signal_name: TheSignal
    value:
      Double: 1
